src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1417 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.Json;
+using Microsoft.IdentityModel.Json.Linq;
+using Microsoft.IdentityModel.JsonWebTokens;
+using Microsoft.IdentityModel.Logging;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.IdentityModel.Protocols.SignedHttpRequest
+{
+    /// <summary>
+    /// A handler designed for creating and validating signed http requests. 
+    /// </summary>
+    /// <remarks>The handler implementation is based on 'A Method for Signing HTTP Requests for OAuth' specification.</remarks>
+    public class SignedHttpRequestHandler
+    {
+        // (https://tools.ietf.org/html/draft-ietf-oauth-signed-http-request-03#section-3.2)
+        // "Encodes the name and value of the header as "name: value" and appends it to the string buffer separated by a newline "\n" character."
+        private readonly string _newlineSeparator = "\n";
+
+        private readonly JsonWebTokenHandler _jwtTokenHandler = new JsonWebTokenHandler();
+        private readonly Uri _baseUriHelper = new Uri("http://localhost", UriKind.Absolute);
+        private readonly HttpClient _defaultHttpClient = new HttpClient();
+
+        #region SignedHttpRequest creation
+        /// <summary>
+        /// Creates a signed http request using the <paramref name="signedHttpRequestDescriptor"/>.
+        /// /// </summary>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <param name="cancellationToken">Propagates notification that operations should be canceled.</param>
+        /// <returns>A signed http request as a JWS in Compact Serialization Format.</returns>
+        public async Task<string> CreateSignedHttpRequestAsync(SignedHttpRequestDescriptor signedHttpRequestDescriptor, CancellationToken cancellationToken)
```

Comment:
Our code would be simpler if when creating a JWS this class focused on building the payload and then calling JsonWebTokenHandler.CreateToken(string, SigningCredentials, IDictionary<string, object>)

see: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/110ced71d96efaaaa4db42eb5f3737b7f31ec2d0/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs#L221

Reply:
I'll refactor the code, it makes sense to utilize JsonWebTokenHandler. The method will be synchronous for now, but we can add another async method when we support async in JsonWebTokenHandler.

Reply:
Addressed with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1260/commits/a1656abf4f653567a5d6e97a49aeed2559d331a5